eclipse-theia/theia
# Introduce lock to prevent parallel task execution

Signed-off-by: Thomas Mäder <tmader@redhat.com>





#### What it does
Fixes https://github.com/eclipse-theia/theia/issues/9806

The idea is to have a lock that prevents interleaved execution of the check whether a task is already running with the actual starting of the task. 
The approach is to delay later execution of tasks instead of ignoring them, because `runTask()` is used in various places that treat failure to start a task as an error.

#### How to test
to reproduce first: 

* Install the extension from the issue: https://github.com/eclipse-theia/theia/files/6922291/prefs-explorer-0.0.1.vsix.zip
* Change the theia source like described here: https://github.com/eclipse-theia/theia/issues/9806#issuecomment-891618281

You should be able to start multiple instances of the same task by clicking on the tasks view.

to verify it's fixed, keep the above changes, but check out the PR branch

You should not be able to create multiple instances of the same task.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
